File: packages/api/src/auth/middleware.js
const { verifyAccessToken } = require("./jwt");

// Rate limiting for auth endpoints
const failedAttempts = new Map(); // IP -> { count, lockedUntil }
const MAX_ATTEMPTS = 5;
const LOCKOUT_MS = 15 * 60 * 1000; // 15 minutes

/** Check if IP is rate limited */
function isRateLimited(ip) {
  const entry = failedAttempts.get(ip);
  if (!entry) return false;
  if (entry.lockedUntil && Date.now() < entry.lockedUntil) return true;
  if (entry.lockedUntil && Date.now() >= entry.lockedUntil) {
    failedAttempts.delete(ip);
    return false;
  }
  return false;
}

/** Record a failed auth attempt */
function recordFailedAttempt(ip) {
  const entry = failedAttempts.get(ip) || { count: 0, lockedUntil: null };
  entry.count++;
  if (entry.count >= MAX_ATTEMPTS) {
    entry.lockedUntil = Date.now() + LOCKOUT_MS;
  }
  failedAttempts.set(ip, entry);
}

/** Clear failed attempts on success */
function clearFailedAttempts(ip) {
  failedAttempts.delete(ip);
}

// Periodically clean up expired lockouts to prevent memory leak
setInterval(() => {
  const now = Date.now();
  for (const [ip, entry] of failedAttempts) {
    if (entry.lockedUntil && now >= entry.lockedUntil) {
      failedAttempts.delete(ip);
    }
  }
}, 60 * 60 * 1000); // Every hour

/** Rate limit middleware for auth routes */
function rateLimitAuth(req, res, next) {
  const ip = req.ip || req.connection.remoteAddress;
  if (isRateLimited(ip)) {
    return res.status(429).json({
      error: "Too many failed attempts. Try again in 15 minutes.",
    });
  }
  req._authIp = ip;
  next();
}

/** JWT authentication middleware */
function requireAuth(req, res, next) {
  const header = req.headers.authorization;
  if (!header || !header.startsWith("Bearer ")) {
    return res.status(401).json({ error: "Missing or invalid Authorization header" });
  }

  const token = header.slice(7);
  try {
    const payload = verifyAccessToken(token);
    req.user = { id: payload.sub, role: payload.role, email: payload.email };
    next();
  } catch (err) {
    if (err.name === "TokenExpiredError") {
      return res.status(401).json({ error: "Token expired" });
    }
    return res.status(401).json({ error: "Invalid token" });
  }
}

/** Role check middleware factory */
function requireRole(...roles) {
  return (req, res, next) => {
    if (!req.user) {
      return res.status(401).json({ error: "Not authenticated" });
    }
    if (!roles.includes(req.user.role)) {
      return res.status(403).json({ error: "Insufficient permissions" });
    }
    next();
  };
}

module.exports = {
  rateLimitAuth,
  recordFailedAttempt,
  clearFailedAttempts,
  requireAuth,
  requireRole,
};

File: packages/windows-agent/auth/middleware.py
import time
from functools import wraps
from flask import request, jsonify, g
from auth.jwt_auth import verify_access_token

# Rate limiting
_failed_attempts = {}  # ip -> {count, locked_until}
MAX_ATTEMPTS = 5
LOCKOUT_SECONDS = 900  # 15 minutes


def is_rate_limited(ip):
    entry = _failed_attempts.get(ip)
    if not entry:
        return False
    if entry.get("locked_until") and time.time() < entry["locked_until"]:
        return True
    if entry.get("locked_until") and time.time() >= entry["locked_until"]:
        del _failed_attempts[ip]
        return False
    return False


def record_failed_attempt(ip):
    entry = _failed_attempts.get(ip, {"count": 0})
    entry["count"] += 1
    if entry["count"] >= MAX_ATTEMPTS:
        entry["locked_until"] = time.time() + LOCKOUT_SECONDS
    _failed_attempts[ip] = entry


def clear_failed_attempts(ip):
    _failed_attempts.pop(ip, None)


def rate_limit_auth(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        ip = request.remote_addr
        if is_rate_limited(ip):
            return jsonify({"error": "Too many failed attempts. Try again in 15 minutes."}), 429
        g.auth_ip = ip
        return f(*args, **kwargs)
    return decorated


def require_auth(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        header = request.headers.get("Authorization", "")
        if not header.startswith("Bearer "):
            return jsonify({"error": "Missing or invalid Authorization header"}), 401
        token = header[7:]
        try:
            payload = verify_access_token(token)
            g.user = {"id": payload["sub"], "role": payload["role"]}
        except Exception as e:
            if "expired" in str(e).lower():
                return jsonify({"error": "Token expired"}), 401
            return jsonify({"error": "Invalid token"}), 401
        return f(*args, **kwargs)
    return decorated


def require_role(*roles):
    def decorator(f):
        @wraps(f)
        def decorated(*args, **kwargs):
            if not hasattr(g, "user"):
                return jsonify({"error": "Not authenticated"}), 401
            if g.user["role"] not in roles:
                return jsonify({"error": "Insufficient permissions"}), 403
            return f(*args, **kwargs)
        return decorated
    return decorator
